File: detect.py
#!/usr/bin/env python2.7
# coding = utf-8

import cv2
import caffe
import numpy as np


FACE_PROB_TH = 0.8


def nms(dets, thresh):
  x1 = dets[:, 0]
  y1 = dets[:, 1]
  x2 = dets[:, 0] + dets[:, 2]
  y2 = dets[:, 1] + dets[:, 3]
  scores = dets[:, 4]

  areas = (x2 - x1 + 1) * (y2 - y1 + 1)
  order = scores.argsort()[::-1]

  keep = []
  while order.size > 0:
    i = order[0]
    keep.append(i)
    xx1 = np.maximum(x1[i], x1[order[1:]])
    yy1 = np.maximum(y1[i], y1[order[1:]])
    xx2 = np.minimum(x2[i], x2[order[1:]])
    yy2 = np.minimum(y2[i], y2[order[1:]])

    w = np.maximum(0.0, xx2 - xx1 + 1)
    h = np.maximum(0.0, yy2 - yy1 + 1)
    inter = w * h
    ovr = inter / (areas[i] + areas[order[1:]] - inter)

    inds = np.where(ovr <= thresh)[0]
    order = order[inds + 1]

  return keep


class Detector(object):
  """Cascaded CNN Face Detector
  """
  def __init__(self):
    """init detector
    """
    self.pnet = caffe.Net('proto/p.prototxt', 'result/p.caffemodel', caffe.TEST)

  def detect(self, img):
    """detect face
    """
    height, width = img.shape[:-1]
    scale_factor = 1
    factor = 1.4
    win_min_size = 20
    face = []
    while min(height, width) >= win_min_size:
      data = img.transpose(2, 0, 1)
      single_face = self.detect_single_scale(data)
      print scale_factor, len(single_face['bbox'])
      for bbox in single_face['bbox']:
        x, y, w, h, s = bbox
        x *= scale_factor
        y *= scale_factor
        w *= scale_factor
        h *= scale_factor
        bbox_new = [x, y, w, h, s]
        face.append(bbox_new)
      height /= factor
      width /= factor
      scale_factor *= factor
      img = cv2.resize(img, (int(height), int(width)))
    face = np.asarray(face)
    keep = nms(face, 0.3)
    face = face[keep, :]
    return face

  def detect_single_scale(self, data):
    """detect single scale face
    """
    # preprocess data
    data = (data.astype(np.float32) - 128) / 128
    height, width = data.shape[1:]
    self.pnet.blobs['data'].reshape(1, 3, height, width)
    self.pnet.blobs['data'].data[...] = data
    output = self.pnet.forward()
    prob = output['face_prob'][0][1]
    bbox_offset = output['face_bbox'][0]
    landmark = output['face_landmark'][0]
    height, width = prob.shape

    face = {'bbox': [], 'landmark': []}
    for i in range(height):
      for j in range(width):
        if prob[i][j] > FACE_PROB_TH:
          x, y, w, h = 2*j, 2*i, 12, 12
          dx, dy, dw, dh = bbox_offset[:, i, j]
          x_ = x + w * dx
          y_ = y + h * dy
          w_ = w * dw
          h_ = h * dh
          bbox = np.array([x, y, w, h, prob[i][j]])
          face['bbox'].append(bbox)
          face['landmark'].append(landmark[:, i, j])
    face['bbox'] = np.asarray(face['bbox'])
    face['landmark'] = np.asarray(face['landmark'])
    return face


if __name__ == '__main__':
  detector = Detector()
  img = cv2.imread('test.jpg', cv2.IMREAD_COLOR)
  # h, w = img.shape[:-1]
  # img = cv2.resize(img, (w/2, h/2))
  face = detector.detect(img)
  for bbox in face:
    x, y, w, h, s = bbox
    cv2.rectangle(img, (int(x), int(y)), (int(x+w), int(y+h)), (0, 0, 255), 1)
    # for x, y in landmark:
    #   cv2.circle(img, (int(x), int(y)), 2, (0, 255, 0), -1)
  cv2.imwrite('res.jpg', img)

File: src/detect.cpp
#include "detect.hpp"
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <caffe/caffe.hpp>

using namespace cv;
using namespace std;

namespace jfda {

vector<FaceBBox> nms(vector<FaceBBox>& bboxes, float overlap = 0.3);

Detector::Detector() {
  pnet = new caffe::Net<float>("../proto/p.prototxt", caffe::TEST);
  pnet->CopyTrainedLayersFromBinaryProto("../result/p.caffemodel");
}

Detector::~Detector() {
  delete pnet;
}

static const float kPNetScoreTh = 0.8;
static const float kRNetScoreTh = 0.8;
static const float kONetScoreTh = 0.8;

vector<FaceBBox> Detector::detect(const Mat& img_) {
  Mat img = img_.clone();
  int height = img.rows;
  int width = img.cols;
  float scale = 1.;
  float factor = 1.4;
  vector<FaceBBox> res;
  boost::shared_ptr<caffe::Blob<float> > input = pnet->blob_by_name("data");
  boost::shared_ptr<caffe::Blob<float> > face_prob = pnet->blob_by_name("face_prob");
  boost::shared_ptr<caffe::Blob<float> > bbox_offset = pnet->blob_by_name("face_bbox");
  TIMER_BEGIN
  int counter = 0;
  while (std::min(img.cols, img.rows) > 20) {
    counter++;
    input->Reshape(1, 3, height, width);
    float* input_data = input->mutable_cpu_data();
    for (int i = 0; i < height; i++) {
      for (int j = 0; j < width; j++) {
        input_data[input->offset(0, 0, i, j)] = static_cast<float>(img.at<cv::Vec3b>(i, j)[0]) / 128 - 1;
        input_data[input->offset(0, 1, i, j)] = static_cast<float>(img.at<cv::Vec3b>(i, j)[1]) / 128 - 1;
        input_data[input->offset(0, 2, i, j)] = static_cast<float>(img.at<cv::Vec3b>(i, j)[2]) / 128 - 1;
      }
    }
    TIMER_BEGIN
    pnet->Forward();
    cout << TIMER_NOW << endl;
    TIMER_END
    int h, w;
    h = face_prob->shape(2);
    w = face_prob->shape(3);
    float* face_prob_data = face_prob->mutable_cpu_data();
    float* bbox_offset_data = bbox_offset->mutable_cpu_data();
    for (int i = 0; i < h; i++) {
      for (int j = 0; j < w; j++) {
        float prob = face_prob_data[face_prob->offset(0, 1, i, j)];
        if (prob > kPNetScoreTh) {
          FaceBBox bbox;
          bbox.x = 2 * j*scale;
          bbox.y = 2 * i*scale;
          bbox.w = bbox.h = 12 * scale;

          bbox.x = bbox.x + bbox.w * bbox_offset_data[bbox_offset->offset(0, 0, i, j)];
          bbox.y = bbox.y + bbox.h * bbox_offset_data[bbox_offset->offset(0, 1, i, j)];
          bbox.w = bbox.w * exp(bbox_offset_data[bbox_offset->offset(0, 2, i, j)]);
          bbox.h = bbox.h * exp(bbox_offset_data[bbox_offset->offset(0, 2, i, j)]);
          bbox.score = prob;
          res.push_back(bbox);
        }
      }
    }

    scale *= factor;
    height = height / factor;
    width = width / factor;
    cv::resize(img, img, cv::Size(width, height));
  }
  cout << counter << endl;
  cout << TIMER_NOW << endl;
  TIMER_END
  return nms(res);
}

vector<FaceBBox> nms(vector<FaceBBox>& bboxes, float overlap) {
  int n = bboxes.size();
  vector<FaceBBox> merged;
  merged.reserve(n);

  for (int i = 0; i < n; i++) {
    for (int j = i + 1; j < n; j++) {
      if (bboxes[i].score < bboxes[j].score) {
        std::swap(bboxes[i], bboxes[j]);
      }
    }
  }

  vector<float> areas(n);
  vector<bool> flag(n, true);
  for (int i = 0; i < n; i++) {
    areas[i] = bboxes[i].w*bboxes[i].h;
  }

  for (int i = 0; i < n; i++) {
    if (flag[i]) {
      merged.push_back(bboxes[i]);
      for (int j = i + 1; j < n; j++) {
        if (flag[i]) {
          float x1 = max(bboxes[i].x, bboxes[j].x);
          float y1 = max(bboxes[i].y, bboxes[j].y);
          float x2 = min(bboxes[i].x + bboxes[i].w, bboxes[j].x + bboxes[j].w);
          float y2 = min(bboxes[i].y + bboxes[i].h, bboxes[j].y + bboxes[j].h);
          float w = max(float(0), x2 - x1);
          float h = max(float(0), y2 - y1);
          float ov = (w*h) / (areas[i] + areas[j] - w*h);
          if (ov > overlap) {
            flag[j] = false;
          }
        }
      }
    }
  }
  return merged;
}

}  // namespace jfda
